helix-term/src/commands.rs

```diff
@@ -3610,6 +3610,39 @@ fn normal_mode(cx: &mut Context) {
 
     doc.mode = Mode::Normal;
 
+    let doc_changes = doc.changes().changes();
+    let text = doc.text().slice(..);
+    let pos = doc.selection(view.id).primary().cursor(text);
+    let mut can_restore_indent = false;
+
+    // Will remove whitespaces when user opens a new line(may additional insert only tabs or spaces)
+    // and goto normal mode.
+    //
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
+    use helix_core::chars::char_is_whitespace;
+    use helix_core::Operation;
+    if let [Operation::Retain(move_pos), Operation::Insert(ref inserted_str), Operation::Retain(_)] =
+        doc_changes
+    {
+        if move_pos + inserted_str.len32() as usize == pos
+            && inserted_str.starts_with('\n')
+            && inserted_str.chars().skip(1).all(char_is_whitespace)
+        {
+            can_restore_indent = true;
+        }
+    }
+
+    if can_restore_indent {
+        let transaction =
+            Transaction::change_by_selection(doc.text(), doc.selection(view.id), |range| {
+                let line_start_pos = text.line_to_char(range.cursor_line(text));
+                (line_start_pos, pos, None)
+            });
+        doc.apply(&transaction, view.id);
+    }
```

Comment:
Let's extract this in a helper function so the code is self contained.

Reply:
Thank you, that make sence a lot :-) Done on that